fabfile.py

```diff
@@ -1,20 +1,161 @@
+    environment
+    """
+    env.stage = stage
+    for option, value in STAGES[env.stage].items():
+        setattr(env, option, value)
+
+
+def ensure_stage(fn):
+    """
+    Decorator to ensure the stage is set
+    """
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        # The require operation will abort if the key stage
+        # is not set in the environment
+        require('stage', provided_by=(staging, production,))
+        return fn(*args, **kwargs)
+    return wrapper
+
+
+@task
+def production():
+    set_stage('production')
+
+
+@task
+def staging():
+    set_stage('staging')
+
+
+@task
+@ensure_stage
+def initialize_server():
+    """
+    Setup an initial deployment on a fresh host.
+    """
+    print 'Setting up the ' + env.stage + ' server'
+    # Sets up a virtualenv directory
+    sr('mkdir', '-p', VENV_DIR)
+    sr('virtualenv', '--python', 'python2', VENV_DIR)
+
+    # Updates current version of ifttt source
+    update_source_repo()
+
+    # Uploads the db and ifttt channel creds to the server
+    upload_config()
+
+    # Updates the virtualenv with latest requirements
+    upgrade_dependencies()
 
 
 @task
-def deploy(*args):
-    with cd(code_dir):
-        sudo('git rev-list HEAD --max-count=1')
-        sudo('git fetch')
-        sudo('git reset --hard origin/master')
-        sudo('webservice uwsgi-python restart')
+@ensure_stage
+def deploy():
+    """
+    Deploys updated code to the web server
+    """
+    print 'Deploying to ' + env.stage
+
+    # Updates current version of ifttt source
+    update_source_repo()
+
+    # Uploads the db and ifttt channel creds to the server
+    upload_config()
+
+    # Updates the virtualenv with latest requirements
+    upgrade_dependencies()
+
+    # Restart ifttt web service
+    restart_ifttt()
+
+
+@ensure_stage
+def update_source_repo():
+    """
+    Update the ifttt source repo
+    """
+    print 'Updating ifttt source repo'
+    with cd(SOURCE_DIR):
+        sr('git', 'fetch', 'origin', env.branch)
+        sr('git', 'reset', '--hard', 'FETCH_HEAD')
+
+
+@ensure_stage
+def upload_config():
+    """
+    Upload config to the remote host
+    """
+    print 'Uploading config files to remote host(s)'
+    put(env.local_config_file, os.path.join(SOURCE_DIR, DEST_CONFIG_FILE),
+        use_sudo=True)
+    sudo("chown www-data:www-data " +
+         os.path.join(SOURCE_DIR, DEST_CONFIG_FILE))
+
+
+@ensure_stage
+def upgrade_dependencies():
+    """
+    Installs upgraded versions of requirements (if applicable)
+    """
+    print 'Upgrading requirements'
+    with cd(VENV_DIR):
+        sr(VENV_DIR + '/bin/pip', 'install', '--upgrade', '-r',
+            os.path.join(SOURCE_DIR, 'requirements.txt'))
+
+
+@task
+@ensure_stage
+def restart_ifttt():
+    """
+    Restarts the ifttt web sersive
+    """
+    print 'Restarting ifttt'
+
+    # Doing uwsgi stop and start due to a known bug in our uwsgi puppet module
```

Comment:
This bug can be fixed by making sure the 'stop-on-term' option is passed to uwsgi.